### ISSUE TITLE
모델 성능 향상

### ISSUE THREAD

**Issue by 17011775** [SEEUNL/Recycle#20]
ResNet 
케라스 창시자에게 배우는 딥러닝 책에서 fit_generator 할 때는 step_per_epoch * batch_size 가 전체가 되도록 만들어 주어야 한다고 한다. 그래서 2000~3000정도가 되도록 설정했다.

batch_size = 2, step per epoch = 1000, epoch = 30일때 
학습 acc = 95 평가 acc = 30

여기서 같은조건으로 data augmentation을 켜고 돌려보니 
학습 acc = 64.5, 평가 acc = 30이 나왔다. 
이처럼  data augmentation을 키니 같은 epoch에서는 underfitting이 되었다.  

https://blog.naver.com/chacagea/221301642313에서 batch size를 설정하는 팁을 보았다.
너무 큰 배치사이즈를 사용하면 메모리가 크고 computing power가 좋아야 한다.
선별한 데이터셋에 노이즈가 있을 수 있어서 너무 작은 배치사이즈는 학습을 방해할 수 있다고 한다. 지금까지는 step_per_epoch * batch_size 가 전체 데이터 사이즈와 비슷하게 맞추기만 하고 배치사이즈를 2,3 같은 작은 수를 사용해왔는데 배치사이즈를 크게 해주는 것이 학습 시간을 단축시키고 결과에도 도움이 되는 케이스가 있다고 한다.  Udacity 강좌에서는 배치사이즈의 추천값을 하나 꼽자면 2의 n승중에 32이고 이것저것 다 해볼 시간이 없다면 32~256 정도의 배치사이즈가 좋다고 한다. epoch는 validation error 가 증가하기 전에 끊어주는 것이 좋다. 

우리 주제와 비슷한 연구를 한 논문(RecycleNet: Intelligent Waste Sorting Using Deep
Neural Networks)을 보니 ResNet을 사용할 때 input size가 224 X 224 였다.  따라서 기존 input size인 24 X 24에서 224 X 224로 바꿔서 실행을 해보았는데 local minimum에 빠진건지 학습이 되지 않아서 원래 사용하던 24 X 24로 학습을 진행했다. 

batch size = 32, step per epoxh = 100, epoch = 100, data augmentation on 일때 
학습 acc = 92 평가 acc = 66 으로 전보다는 괜찮은 결과가 나왔다. 
![KakaoTalk_20190613_012349409](https://user-images.githubusercontent.com/48524429/59368853-0e7b5100-8d7a-11e9-9747-4b8e9534bae1.jpg)

epoch를 늘려도 될 것 같아서 
batch size = 32, step per epoxh = 100, epoch = 130, data augmentation on에 드롭아웃도 추가
학습 acc = 87 평가 acc = 62 가 나왔다. 

오버피팅이 계속 있는것 같아서 드롭아웃을 추가해 보았는데 드롭아웃을 추가하니 학습 속도가 느려지고 epoch를 늘렸지만 오히려 학습 acc는 늘리기 전보다 낮았다. 
오버피팅을 방지하는 방법중 하나라서 사용해봤는데 느려지기만 하고 효과가 있는지 모르겠다.
 https://blog.naver.com/dkdlel531/221480376452에서 보니 요즘에는 dropout대신 batch normalization을 많이 쓴다고 한다. 물론 같이써도 되는데 그렇게 잘 안한다. ResNet 학습 코드에 중간 중간 Batch normalization을 사용해주고 dropout을 사용해봐도 효과가 거의 없는것 같아 굳이 사용하지 않기로 했다. 

또다른 regularization 방법인 L1, L2 정규화를 사용해봤다. 
L1, L2를 같이 사용했을 때와 L1 정규화만 사용했을 때 왜인지는 모르겠지만 cost값이 줄지 않고 local minima에 빠졌다.
하지만 L2 정규화만 사용했을 때는 학습이 제대로 됬다.

batch size = 32, step per epoxh = 100, epoch = 130, data augmentation on, L2 정규화
학습 acc = 92, 평가 acc = 74
![KakaoTalk_20190613_013923066](https://user-images.githubusercontent.com/48524429/59369787-1f2cc680-8d7c-11e9-913f-4cd39b12fd68.jpg)
validation accuracy 그래프가 굉장히 꾸불거리지만 평가  accuracy가 나쁘지 않아서 
같은 조건에서 batch size의 영향을 알아보기 위해 step_per_epoch * batch_size 가 전체 데이터 사이즈와 비슷하게 맞추고 batch size와 step per epoch만 바꿔서 실행해 보았다.
위와 같은 조건 batch size= 64, step per epoch = 50 일때 
학습 acc  = 94 평가 acc = 66
![KakaoTalk_20190613_014616055](https://user-images.githubusercontent.com/48524429/59370309-36b87f00-8d7d-11e9-94b7-741581455ab7.jpg)

위와 같은 조건 batch size= 128, step per epoch = 25 일때
학습 acc  = 93 평가 acc = 60
![KakaoTalk_20190613_014629437](https://user-images.githubusercontent.com/48524429/59370322-3c15c980-8d7d-11e9-844d-3754b2f2a91e.jpg)

배치사이즈를 늘릴 수록 validation 그래프의 꾸불거림이 줄어들었다. 
Jay Alammar님이 괜히 추천해주신게 아닌지 32일 때 가장 평가 acc가 높았다
지금까지 알고있는 오버피팅 방지 방법( L1,L2 정규화, 데이터 augmentation, 드롭아웃, batch normalization, 데이터 추가 수집 등)  왠만한 것들은 적용시켜 보았지만 평가 정확도가 잘나와야 6~70퍼센트 정도라서 어떻게 늘릴 수 있을지 고민을 해봐야 할 것 같다.
논문에서 같은 모델, 같은 조건에서 optimizer를 adam 대신 adadelta를 사용하니까 미미하게 결과가 좋아졌지만 거의 별 효과는 없는 것 같다(1프로정도 향상).

**Comment by unizard** [comment]
@17011775  클래스별 정확도를 좀 보여줄래요? Confusion matrix 좀 뽑아줘
